controllers/nodefeaturediscovery_finalizers.go

```diff
@@ -0,0 +1,411 @@
+				return ctrl.Result{Requeue: false}, nil
+			}
+			klog.Info("Finalizer was found, but removing it was unsuccessful. Requeueing deletion request.")
+			return ctrl.Result{}, nil
+		}
+
+		klog.Info("Finalizer was found and successfully removed.")
+		return ctrl.Result{Requeue: false}, nil
+	}
+
+	klog.Info("Finalizer does not exist, but resource deletion succesful.")
+	return ctrl.Result{Requeue: false}, nil
+}
+
+// addFinalizer adds a finalizer to the NFD Operator instance
+func (r *NodeFeatureDiscoveryReconciler) addFinalizer(ctx context.Context, instance *nfdv1.NodeFeatureDiscovery, finalizer string) (ctrl.Result, error) {
+	// Add the defined finalizer as a finalizer to the instance if it does not exist
+	instance.Finalizers = append(instance.Finalizers, finalizer)
+	instance.Status.Conditions = r.getProgressingConditions("DeploymentStarting", "Deployment is starting")
+	if err := r.Update(ctx, instance); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// we exit reconcile loop because we will have additional update reconcile
+	return ctrl.Result{Requeue: false}, nil
+}
+
+// hasFinalizer determines if the operator instance has a specific
+// finalizer value, which is defined by the parameter 'finalizer'
+func (r *NodeFeatureDiscoveryReconciler) hasFinalizer(instance *nfdv1.NodeFeatureDiscovery, finalizer string) bool {
+	if len(instance.Finalizers) == 0 {
+		return false
+	}
+
+	// The instance will have a list of finalizers under its
+	// `metav1.ObjectMeta` reference
+	for _, f := range instance.Finalizers {
+
+		// If the current finalizer in the list matches the
+		// 'finalizer' parameter, then the operator does have
+		// the desired finalizer, so return "true"
+		if f == finalizer {
+			return true
+		}
+	}
+
+	// Return false, as the finalizer was not found in the list.
+	return false
+}
+
+// removeFinalizer removes a finalizer from the operator's instance
+func (r *NodeFeatureDiscoveryReconciler) removeFinalizer(instance *nfdv1.NodeFeatureDiscovery, finalizer string) {
+	// 'finalizers' will contain a list of all the finalizers for
+	// the NFD operator instance, except for the finalizer that
+	// is being removed. (The finalizer to remove is defined with
+	// this function's parameter 'finalizer'.)
+	var finalizers []string
+
+	// The instance will have a list of finalizers under its
+	// `metav1.ObjectMeta` reference
+	for _, f := range instance.Finalizers {
+
+		// If the current finalizer in the list matches the
+		// 'finalizer' parameter, then we want to remove it.
+		// However, rather than delete from the list, it is
+		// more efficient to just create a new list and set
+		// the 'Finalizers' attribute to that new list. Thus,
+		// this part of the loop skips the addition of the
+		// finalizer we want to remove.
+		if f == finalizer {
+			continue
+		}
+		finalizers = append(finalizers, f)
+	}
+
+	// Update the 'Finalizers' attribute to point to the newly
+	// updated list.
+	instance.Finalizers = finalizers
```

Comment:
nit: trim down the excessive comments

```suggestion
	var finalizers []string

	for _, f := range instance.Finalizers {
		if f != finalizer {
			finalizers = append(finalizers, f)
		}
	}
	
	instance.Finalizers = finalizers
```